src/Solvers/preconditioned_conjugate_gradient_solver.jl

```diff
@@ -1,109 +1,128 @@
-    )
-
-   return PreconditionedConjugateGradientSolver(arch, settings)
+See `solve!` for more information about the preconditioned conjugate-gradient algorithm.
+"""
+function PreconditionedConjugateGradientSolver(linear_operation;
+                                               template_field::AbstractField,
+                                               maximum_iterations = prod(size(template_field.grid)),
+                                               tolerance = 1e-13,
+                                               precondition = nothing)
+
+    arch = architecture(template_field)
+    grid = template_field.grid
+
+    # Create work arrays for solver
+    linear_operator_product = similar(template_field) # A*xᵢ = qᵢ
+    search_direction = similar(template_field) # pᵢ
+            residual = similar(template_field) # rᵢ
+
+    # Either nothing (no precondition) or P*xᵢ = zᵢ
+    precondition_product = initialize_precondition_product(precondition, template_field)
+
+    return PreconditionedConjugateGradientSolver(arch,
+                                                 grid,
+                                                 linear_operation,
+                                                 tolerance,
+                                                 maximum_iterations,
+                                                 0,
+                                                 0.0,
+                                                 linear_operator_product,
+                                                 search_direction,
+                                                 residual,
+                                                 precondition,
+                                                 precondition_product)
 end
 
-@kernel function compute_residual!(r, RHS, A)
-    i, j, k = @index(Global, NTuple)
-    @inbounds r[i, j, k] = RHS[i, j, k] - A[i, j, k]
+function Statistics.norm(a::AbstractField)
+    ii = interior_parent_indices(location(a, 1), topology(a.grid, 1), a.grid.Nx, a.grid.Hx)
+    ji = interior_parent_indices(location(a, 2), topology(a.grid, 2), a.grid.Ny, a.grid.Hy)
+    ki = interior_parent_indices(location(a, 3), topology(a.grid, 3), a.grid.Nz, a.grid.Hz)
+    return sqrt(mapreduce(x -> x * x, +, view(parent(a), ii, ji, ki)))
 end
 
-function quick_launch!(arch, grid, kernel!, args...)
-    event = launch!(arch, grid, :xyz, kernel!, args...; dependencies=Event(device(arch)))
-    wait(device(arch), event)
-    return nothing
+function Statistics.dot(a::AbstractField, b::AbstractField)
+    ii = interior_parent_indices(location(a, 1), topology(a.grid, 1), a.grid.Nx, a.grid.Hx)
+    ji = interior_parent_indices(location(a, 2), topology(a.grid, 2), a.grid.Ny, a.grid.Hy)
+    ki = interior_parent_indices(location(a, 3), topology(a.grid, 3), a.grid.Nz, a.grid.Hz)
+    return mapreduce((x, y) -> x * y, +,
+                     view(parent(a), ii, ji, ki),
+                     view(parent(b), ii, ji, ki))
 end
```

Comment:
Should we move these to `Oceananigans.Field`? They seem useful as general utils.

Reply:
We can do that now or wait for the refactoring that will have to occur when we extend these to work in parallel and for the cubed sphere?

Reply:
Doing that later, as part of cubesphere "faceificaton" makes sense to me.